### PR TITLE
New classes for SampleBasedMetric, custom signatures for plots and summarizer dependence. Also adds automated unit tests

### PR DESCRIPTION
**Pull request by CompiledAtBirth**
Hello, below is the list of changes in this PR:

- PosteriorCoverage and PosteriorSamples classes in [ili.validation.metrics](https://github.com/maho3/ltu-ili/blob/nico_50_56_71_73/ili/validation/metrics.py). The former combines Rank Statistics and TARP [issue 56](https://github.com/maho3/ltu-ili/issues/56) and the latter allows to save posterior samples for ulterior use [issue 50](https://github.com/maho3/ltu-ili/issues/50).
- Dataset class added in [ili.dataloaders.loaders](https://github.com/maho3/ltu-ili/blob/nico_50_56_71_73/ili/dataloaders/loaders.py) to remove, at least for now, ili-summarizer dependency from [issue 58](https://github.com/maho3/ltu-ili/issues/78).
- ensemble_mode argument added in [ili.validation.runner](https://github.com/maho3/ltu-ili/blob/nico_50_56_71_73/ili/validation/runner.py). The default behavior for a validation runner after loading an ensemble posterior (eg with a trained MAf and a trained MDN) is to compute the metrics for the [NeuralPosteriorEnsemble](https://github.com/mackelab/sbi/blob/main/sbi/utils/posterior_ensemble.py) sbi instance. In this case, the posteriors in the ensemble have their val_log_prob on validation data as weights. Alternatively, we can now compute the metrics for each trained posterior in the ensemble using ensemble_mode = False. This has been added in the example [toy_sbi](https://github.com/maho3/ltu-ili/blob/nico_50_56_71_73/examples/configs/val/toy_sbi.yaml). I have not dealt with the pydelfi backend.
- added signatures and str_save in [ili.inference.runner_sbi](https://github.com/maho3/ltu-ili/blob/nico_50_56_71_73/ili/inference/runner_sbi.py), and signatures for [ili.validation.runner](https://github.com/maho3/ltu-ili/blob/nico_50_56_71_73/ili/validation/runner.py). For ease of usage, these signatures are passed as arguments to call functions in [ili.validation.metrics](https://github.com/maho3/ltu-ili/blob/nico_50_56_71_73/ili/validation/metrics.py) instead of being attributes.
- In addition to the [toy_sbi*.py](https://github.com/maho3/ltu-ili/tree/nico_50_56_71_73/examples) examples, I also modified the [tutorial](https://github.com/maho3/ltu-ili/blob/nico_50_56_71_73/notebooks/tutorial.ipynb) notebook so that it works with the PosteriorCoverage class. I have not taken the time to modify nor rerun the third part of the tutorial using CAMELS data.

Please share your thoughts, especially about the ensemble_mode feature which I took the initiative to implement to deal with the case where we want to compare different models (MAF, MDN...) in the same sbi runner instance.

### CODE REVIEW

**Comment by maho3** [comment]
Overall, this PR is huge. Really great work. This really cleans up many of the details that we needed fixed a while ago. Thanks Nicolas

The `ensemble_mode` thing is a really good idea. That will be really useful for checking multiple architectures at once. People may just end up using `ensemble_mode=True` most of the time, but the option will be very useful for trial-and-error.

I need to do a thorough review this week of all these changes. Here are some checklist for myself:
- [ ] The ili_summarizer `Dataset` class might make  more sense sitting in `ili/utils`, since its not a dataloader  itself.
- [ ] Integrate or catch the `ensemble_mode` parameter in the pydelfi version
- [ ] Test PosteriorCoverage and PosteriorSamples
- [ ] Test signature functionality.
- [ ] Run full tutorial again
- [ ] Update INSTALL.md and related docs, removing ili-summarizer dependency.
- [ ] Add signature and ensemble_mode explanations to CONFIG.md

**Comment by CompiledAtBirth** [comment]
Thank you Matt for the comments and the checklist.
@DeaglanBartlett, since issue #68 involves a lot of work for the test suite, do you want to link the issue here as well before thinking about merging? I hope the fact that the PosteriorCoverage class replaces the rank stats and TARP metrics, among other things, is not a pain to rewrite existing validation scripts.

**Comment by DeaglanBartlett** [comment]
@CompiledAtBirth yes it makes sense to wait for this PR to be approved before making a PR for #68 so that the changes made here can be added to the test suite.

**Comment by CompiledAtBirth** [comment]
Thank you Deaglan for all the changes. I just added the possibility to use existing sims at first round for multiround inference (for working example 2).

File: ili/validation/runner.py [hunk 1]
```diff
@@ -63,13 +69,26 @@ def from_config(cls, config_path) -> "ValidationRunner":
...
+            ensemble_mode = config["ensemble_mode"]
+        else:
+            ensemble_mode = True
 
+        print("Number of posteriors in the ensemble is %i."%posterior.num_components)
+        if ensemble_mode:
+            print("Metrics are computed for the ensemble posterior estimate.")
+        else:
+            print("Metrics are computed for each posterior in the ensemble.")
+        #print(vars(posterior.posteriors[0]))
```
Comment (maho3) [inline ili/validation/runner.py:90]: Remove

File: ili/validation/metrics.py [hunk 2]
```diff
@@ -344,38 +418,73 @@ def __call__(
...
+        # Initialize posterior samples array ; careful about size consistency given backend
         if self.sample_method == "emcee":
             P = sampler.num_chains*self.num_samples
         else:
             P = self.num_samples
-        posterior_samples = np.zeros(
-            (P, x.shape[0], theta.shape[1]))
-        for ii in tqdm.tqdm(range(x.shape[0])):
+            
+        # shape = (num_samples, ntest = x.shape[0], ndim)
```
Comment (maho3) [inline ili/validation/metrics.py:445]: Remove

File: ili/validation/metrics.py [hunk 3]
```diff
@@ -56,6 +59,7 @@ def __init__(
         sample_method: str = 'emcee',
         sample_params: dict = {},
         labels: Optional[List[str]] = None,
+        #signature: Optional[str] = "",
```
Comment (maho3) [inline ili/validation/metrics.py:62]: Remove

File: ili/validation/metrics.py [hunk 4]
```diff
@@ -70,16 +74,79 @@ def _build_sampler(self, posterior) -> _BaseSampler:
             if self.backend != 'sbi':
                 raise ValueError(
                     'Pyro backend is only available for sbi posteriors')
+                    
             # check if DirectPosterior is available
+            # First case: we have a NeuralPosteriorEnsemble instance
             if isinstance(posterior, NeuralPosteriorEnsemble):
+                # We only need to check the first element (one model class for a posterior ensemble)
                 if isinstance(posterior.posteriors[0], DirectPosterior):
```
Comment (maho3) [inline ili/validation/metrics.py:82]: We should probably remove this. Sometimes, people may want to use mcmc for SNPE methods...

File: ili/validation/metrics.py [hunk 5]
```diff
@@ -143,23 +211,35 @@ def __call__(
...
+        strFig = signature + "plot_single_posterior.jpg"
+        g.savefig(self.output_path / strFig,
                   dpi=200, bbox_inches='tight')
-
-
-class PlotRankStatistics(_SampleBasedMetric):
+        
+class PosteriorCoverage(_SampleBasedMetric):
+    
     """Plot rank histogram, posterior coverage, and true-pred diagnostics
```
Comment (maho3) [inline ili/validation/metrics.py:220]: Cleanup docstring

File: ili/validation/metrics.py [hunk 6]
```diff
@@ -283,50 +373,34 @@ def _plot_predictions(self, trues, mus, stds):
...
+ 
     def __call__(
         self,
         posterior: ModelClass,
         x: np.array,
         theta: np.array,
         x_obs: Optional[np.array] = None,
         theta_obs: Optional[np.array] = None,
+        signature: Optional[str] = "",
+        plot_list: Optional[list] = ["coverage", "histogram", "predictions", "TARP"], #plot each metric by default
```
Comment (maho3) [inline ili/validation/metrics.py:403]: let's make everything lowercase

File: ili/inference/runner_sbi.py [hunk 7]
```diff
@@ -316,14 +370,24 @@ def __call__(self, loader: _BaseLoader):
...
             )
-
-            with open(self.output_path / f"posterior_{rnd}.pkl", "wb") as f:
-                pickle.dump(posterior, f)
-            proposal = posterior.set_default_x(x_obs)
+            
+            if self.signatures is None:
+                self.signatures = [""]*len(posteriors)
+                
+            print(self.signatures)
```
Comment (maho3) [inline ili/inference/runner_sbi.py:377]: Change to logging?

File: ili/inference/runner_sbi.py [hunk 8]
```diff
@@ -282,13 +315,34 @@ def __call__(self, loader: _BaseLoader):
...
 
+        # loader has x and theta attributes, both default values are None
+        # Even in multiround inference, we can take advantage of prerun simulation-parameter pairs
+        x = loader.get_all_data()
+        theta = loader.get_all_parameters()
+        if x is not None and theta is not None:
+            theta = torch.Tensor(theta).to(self.device)
+            x = torch.Tensor(x).to(self.device)
+            prerun_sims = True
+            print("The first round of inference will use existing sims from the loader.")
```
Comment (maho3) [inline ili/inference/runner_sbi.py:326]: Change to logging

**Review by maho3 — COMMENTED** [review]
Mostly comments on typos/cleanup. No functionality complaints. I will soon push a commit to clean all these things

File: devNotes.txt
[the remark this answers is not in the document]
Comment (maho3) [reply devNotes.txt:1]: Deleted.

File: ili/dataloaders/loaders.py [hunk 9]
```diff
@@ -9,14 +9,18 @@
 import numpy as np
 import json
 import pandas as pd
-from summarizer.dataset import Dataset
+from typing import Dict
+import xarray as xr
```
[the remark this answers is not in the document]
Comment (maho3) [reply ili/dataloaders/loaders.py:13]: Removed.

File: ili/inference/runner_sbi.py [hunk 10]
```diff
@@ -204,10 +227,11 @@ def __call__(self, loader: _BaseLoader, seed: int = None):
             loader (_BaseLoader): dataloader with stored summary-parameter pairs
             seed (int): torch seed for reproducibility
         """
-
         t0 = time.time()
         x = torch.Tensor(loader.get_all_data()).to(self.device)
         theta = torch.Tensor(loader.get_all_parameters()).to(self.device)
+        
+        print("INFERENCE CLASS For MODELS: %s"%self.class_name)
```
[the remark this answers is not in the document]
Comment (maho3) [reply ili/inference/runner_sbi.py:234]: Done.

File: ili/inference/runner_sbi.py [hunk 11]
```diff
@@ -223,18 +247,18 @@ def __call__(self, loader: _BaseLoader, seed: int = None):
...
             # setup training class
             if "SNPE" in self.class_name:
                 model = self._setup_SNPE(net, theta, x)
             elif "SNLE" in self.class_name or "MNLE" in self.class_name:
                 model = self._setup_SNLE(net, theta, x)
             elif "SNRE" in self.class_name or "BNRE" in self.class_name:
                 model = self._setup_SNRE(net, theta, x)
-
+            
+            #print(dir(model))
```
[the remark this answers is not in the document]
Comment (maho3) [reply ili/inference/runner_sbi.py:258]: Done.

File: ili/inference/runner_sbi.py [hunk 12]
```diff
@@ -245,16 +269,25 @@ def __call__(self, loader: _BaseLoader, seed: int = None):
         ).to(self.device)
         posterior = NeuralPosteriorEnsemble(
             posteriors=posteriors, weights=weights)
+        
+        if self.signatures is None:
+            self.signatures = [""]*len(summaries)
+        
+        # Instanciate custom class with NPE and signature
```
[the remark this answers is not in the document]
Comment (maho3) [reply ili/inference/runner_sbi.py:276]: Fixed.

File: ili/validation/runner.py (hunk 1 above)
Comment (maho3) [reply ili/validation/runner.py:90]: Done.

File: ili/validation/metrics.py (hunk 2 above)
Comment (maho3) [reply ili/validation/metrics.py:445]: Done.

File: ili/validation/metrics.py (hunk 3 above)
Comment (maho3) [reply ili/validation/metrics.py:62]: Done.

File: ili/inference/runner_sbi.py (hunk 7 above)
Comment (maho3) [reply ili/inference/runner_sbi.py:377]: Done.

File: examples/configs/infer/toy_sbi_SNPE.yaml [hunk 13]
```diff
@@ -9,13 +9,15 @@ prior:
 model:
   module: 'sbi.inference'
   class: 'SNPE'
+  str_save: 'new_PR_'
```
[the remark this answers is not in the document]
Comment (maho3) [reply examples/configs/infer/toy_sbi_SNPE.yaml:12]: Changed to toy_SNPE_ and such...

File: ili/validation/metrics.py (hunk 5 above)
Comment (maho3) [reply ili/validation/metrics.py:220]: Done.

File: ili/validation/metrics.py (hunk 4 above)
Comment (maho3) [reply ili/validation/metrics.py]: Done. Now, we only use DirectPosteriors if you specify `'sample_method': 'direct'`

**Comment by maho3** [comment]
I think this PR is nearly good to go. I've gone around, tested things, and cleaned up code without changing much functionality. I've also rerun the tutorial.ipynb notebook and updated the documentation.

However, there's two changes I'd like to recommend, but first want @CompiledAtBirth 's approval on before making them:

1) I think the meaning of `str_save` is a little obscure at first glance. I would recommend changing it to just `name` or something.

2) Instead of wrapping the `NeuralPosteriorEnsemble` in another layer of abstraction as in [the `PosteriorEnsemble` class](https://github.com/maho3/ltu-ili/blob/b58738b85b19723d71501150a382717cdbcf3d1d/ili/inference/runner_sbi.py#L28C2-L28C2), I would recommend that we just directly assign signatures as an attribute of sbi's `NeuralPosteriorEnsemble`.

For example, in its usage for saving [in this line](https://github.com/maho3/ltu-ili/blob/b58738b85b19723d71501150a382717cdbcf3d1d/ili/inference/runner_sbi.py#L389), we can just do something like:
```python
posterior = NeuralPosteriorEnsemble(
    posteriors=posteriors,
    weights=val_logprob
)
...
posterior.signatures = signatures
```
Then, when we pickle NeuralPosteriorEnsemble later, signatures should be saved with it.

This way, we don't have to do the awkward thing of e.g. `posterior_ensemble.ensemble.sample` to access the useful `NeuralPosteriorEnsemble` functions, as in the jupyter tutorial. This will also retain consistency with the pydelfi interface.

Just some recommendations, but I'm curious what you think. After these are addressed and changes are propogated through the code (including rerunning tutorial), I think I'm ready to approve the PR.

**Comment by CompiledAtBirth** [comment]
@maho3 super, thank you for the thoughtful review and cleaning. I agree on both points which make usage and future code modifications more straightforward. Special mention to 2), I did not know that one can add attributes dynamically to a class instance in Python :)

**Review by maho3 — APPROVED** [review]

**Pull request merged**